### ISSUE TITLE
UnicodeDecodeError in addLogfile (python 3.7.3 worker)

### ISSUE THREAD

**Issue by vbraun** [buildbot/buildbot#4744]
One buildbot worker now dies in the middle of a build:
~~~~
2019-05-01 09:14:22+0200 [-] unexpected error in processEnded
	Traceback (most recent call last):
	  File "/usr/lib64/python3.7/site-packages/twisted/internet/process.py", line 985, in maybeCallProcessEnded
	    self.reapProcess()
	  File "/usr/lib64/python3.7/site-packages/twisted/internet/process.py", line 350, in reapProcess
	    self.processEnded(status)
	  File "/usr/lib64/python3.7/site-packages/twisted/internet/_baseprocess.py", line 52, in processEnded
	    self.maybeCallProcessEnded()
	  File "/usr/lib64/python3.7/site-packages/twisted/internet/process.py", line 987, in maybeCallProcessEnded
	    _BaseProcess.maybeCallProcessEnded(self)
	--- <exception caught here> ---
	  File "/usr/lib64/python3.7/site-packages/twisted/internet/_baseprocess.py", line 64, in maybeCallProcessEnded
	    proto.processEnded(Failure(reason))
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/runprocess.py", line 252, in processEnded
	    self.command.finished(sig, rc)
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/runprocess.py", line 784, in finished
	    w.stop()
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/runprocess.py", line 131, in stop
	    self.poll()
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/runprocess.py", line 166, in poll
	    self.command.addLogfile(self.name, data)
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/runprocess.py", line 773, in addLogfile
	    self._addToBuffers(('log', name), data)
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/runprocess.py", line 749, in _addToBuffers
	    self._sendBuffers()
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/runprocess.py", line 731, in _sendBuffers
	    self._sendMessage(msg)
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/runprocess.py", line 677, in _sendMessage
	    msg = self._collapseMsg(msg)
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/runprocess.py", line 663, in _collapseMsg
	    m = bytes2unicode(m, self.builder.unicode_encoding)
	  File "/usr/lib/python3.7/site-packages/buildbot_worker/compat.py", line 87, in bytes2unicode
	    return text_type(x, encoding, errors)
	builtins.UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 9872: invalid start byte
~~~~
The worker is fedora's buildbot-1.8.1 (python 3):
~~~~
[buildslave-sage@volker ~]$ rpm -qa | grep -i buildbot
buildbot-www-1.8.1-1.fc29.noarch
buildbot-master-1.8.1-1.fc29.noarch
buildbot-1.8.1-1.fc29.noarch
buildbot-worker-1.8.1-1.fc29.noarch
~~~~
We do have some older successful builds with buildbot-worker-1.8.1, but python 3.7.3 was installed just after the last successful build

**Comment by vbraun** [comment]
I added some logging and the troublesome value is 
~~~~
b'/usr/bin/cat:      260 ms (186 ms GC) and 62.5KB allocated for streams.tst\ntesting: /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/var/tmp/sag\\\ne/build/gap-4.10.1.p0/src/tst/testinstall/strings.tst\n########> Diff in /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/va\\\nr/tmp/sage/build/gap-4.10.1.p0/src/tst/testinstall/strings.tst:44\n# Input is:\nx:="\\0xFF";\n# Expected output:\n"\\377"\n# But found:\n"\xff"\n########\n########> Diff in /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/va\\\nr/tmp/sage/build/gap-4.10.1.p0/src/tst/testinstall/strings.tst:48\n# Input is:\nx:="A string with \\0xFF Hex stuff \\0x42 in it";\n# Expected output:\n"A string with \\377 Hex stuff B in it"\n# But found:\n"A string with \xff Hex stuff B in it"\n########\n########> Diff in /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/va\\\nr/tmp/sage/build/gap-4.10.1.p0/src/tst/testinstall/strings.tst:189\n# Input is:\nList("abcxyzABCXYZ\\n\\t !019?\\377\\000$", LowercaseChar);\n# Expected output:\n"abcxyzabcxyz\\n\\t !019?\\377\\000$"\n# But found:\n"abcxyzabcxyz\\n\\t !019?\xff\\000$"\n########\n########> Diff in /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/va\\\nr/tmp/sage/build/gap-4.10.1.p0/src/tst/testinstall/strings.tst:191\n# Input is:\nList("abcxyzABCXYZ\\n\\t !019?\\377\\000$", UppercaseChar);\n# Expected output:\n"ABCXYZABCXYZ\\n\\t !019?\\377\\000$"\n# But found:\n"ABCXYZABCXYZ\\n\\t !019?\xff\\000$"\n########\n########> Diff in /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/va\\\nr/tmp/sage/build/gap-4.10.1.p0/src/tst/testinstall/strings.tst:193\n# Input is:\nUppercaseString("abcxyzABCXYZ\\n\\t !019?\\377\\000$");\n# Expected output:\n"ABCXYZABCXYZ\\n\\t !019?\\377\\000$"\n# But found:\n"ABCXYZABCXYZ\\n\\t !019?\xff\\000$"\n########\n########> Diff in /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/va\\\nr/tmp/sage/build/gap-4.10.1.p0/src/tst/testinstall/strings.tst:195\n# Input is:\nLowercaseString("abcxyzABCXYZ\\n\\t !019?\\377\\000$");\n# Expected output:\n"abcxyzabcxyz\\n\\t !019?\\377\\000$"\n# But found:\n"abcxyzabcxyz\\n\\t !019?\xff\\000$"\n########\n     780 ms (291 ms GC) and 4.79MB allocated for strings.tst\ntesting: /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/var/tmp/sag\\\ne/build/gap-4.10.1.p0/src/tst/testinstall/switch.tst\n     287 ms (196 ms GC) and 44.7KB allocated for switch.tst\ntesting: /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/var/tmp/sag\\\ne/build/gap-4.10.1.p0/src/tst/testinstall/switch_obj.tst\n     715 ms (508 ms GC) and 57.4KB allocated for switch_obj.tst\ntesting: /mnt/disk/home/buildslave-sage/slave/sage_git/build/local/var/tmp/sag\\\ne/build/gap-4.10.1.p0/src/tst/testinstall/sylowhall.tst\n
~~~~
so a buildstep produces output that is not valid utf-8. Obviously that should be handled without crashing the build worker. The usual choice is either 1) throw away the error or 2) encode it as utf8 surrogates.

**Comment by vbraun** [comment]
I couldn't reproduce this in a unit test, so might be fixed in master. See #4748

**Comment by vbraun** [comment]
I tried buildbot-worker 2.2.0 and the UnicodeDecodeError is still in the log but the worker doesn't die:
~~~~
2019-05-04 13:39:56+0200 [-] Unhandled Error
	Traceback (most recent call last):
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/python/log.py", line 103, in callWithLogger
	    return callWithContext({"system": lp}, func, *args, **kw)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/python/log.py", line 86, in callWithContext
	    return context.call({ILogContext: newCtx}, func, *args, **kw)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/python/context.py", line 122, in callWithContext
	    return self.currentContext().callWithContext(ctx, func, *args, **kw)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/python/context.py", line 85, in callWithContext
	    return func(*args,**kw)
	--- <exception caught here> ---
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/internet/posixbase.py", line 614, in _doReadOrWrite
	    why = selectable.doRead()
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/internet/process.py", line 291, in doRead
	    return fdesc.readFromFD(self.fd, self.dataReceived)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/internet/fdesc.py", line 94, in readFromFD
	    callback(output)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/internet/process.py", line 295, in dataReceived
	    self.proc.childDataReceived(self.name, data)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/internet/process.py", line 961, in childDataReceived
	    self.proto.childDataReceived(name, data)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/twisted/internet/protocol.py", line 640, in childDataReceived
	    self.outReceived(data)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/buildbot_worker/runprocess.py", line 223, in outReceived
	    self.command.addStdout(decodedData)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/buildbot_worker/runprocess.py", line 756, in addStdout
	    self._addToBuffers('stdout', data)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/buildbot_worker/runprocess.py", line 749, in _addToBuffers
	    self._sendBuffers()
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/buildbot_worker/runprocess.py", line 707, in _sendBuffers
	    self._sendMessage(msg)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/buildbot_worker/runprocess.py", line 677, in _sendMessage
	    msg = self._collapseMsg(msg)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/buildbot_worker/runprocess.py", line 663, in _collapseMsg
	    m = bytes2unicode(m, self.builder.unicode_encoding)
	  File "/mnt/disk/home/buildslave-sage/local/lib64/python3.7/site-packages/buildbot_worker/compat.py", line 87, in bytes2unicode
	    return text_type(x, encoding, errors)
	builtins.UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 210: invalid start byte
~~~~

**Comment by p12tic** [comment]
Looks like the bug has been fixed in #4759.
